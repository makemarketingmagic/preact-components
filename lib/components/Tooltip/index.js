'use strict';

exports.__esModule = true;
exports.default = undefined;

var _preact = require('preact');

var _Tooltip = require('./Tooltip.less');

var _Tooltip2 = _interopRequireDefault(_Tooltip);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _CrossIcon = require('../icons/CrossIcon');

var _CrossIcon2 = _interopRequireDefault(_CrossIcon);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Tooltip = function (_Component) {
    _inherits(Tooltip, _Component);

    function Tooltip(props) {
        _classCallCheck(this, Tooltip);

        var _this = _possibleConstructorReturn(this, _Component.call(this, props));

        _this.toggleTooltip = function () {
            _this.setState({ visible: !_this.state.visible });
        };

        _this.state = {
            visible: props.visible || true
        };
        return _this;
    }

    Tooltip.prototype.render = function render() {
        var _props = this.props,
            name = _props.name,
            position = _props.position,
            message = _props.message,
            color = _props.color,
            _props$closable = _props.closable,
            closable = _props$closable === undefined ? true : _props$closable;

        return (0, _preact.h)(
            'div',
            { 'class': (0, _classnames2.default)(_Tooltip2.default.tooltipContainer, 'notranslate', this.state.visible ? _Tooltip2.default.tooltipContainerOpen : _Tooltip2.default.tooltipContainerClosed) },
            (0, _preact.h)(
                'div',
                { 'class': _Tooltip2.default.tooltip },
                (0, _preact.h)(
                    'div',
                    { 'class': _Tooltip2.default.tooltipContents },
                    (0, _preact.h)(
                        'div',
                        null,
                        (0, _preact.h)(
                            'div',
                            { 'class': _Tooltip2.default.tooltipName },
                            name
                        ),
                        (0, _preact.h)(
                            'div',
                            { className: (0, _classnames2.default)(_Tooltip2.default.tooltipPosition, _Tooltip2.default['tooltipPosition--' + color]) },
                            position
                        ),
                        (0, _preact.h)(
                            'div',
                            { 'class': _Tooltip2.default.tooltipMessage },
                            message
                        )
                    ),
                    closable && (0, _preact.h)(
                        'div',
                        { onClick: this.toggleTooltip, 'class': _Tooltip2.default.closeButton },
                        (0, _preact.h)(_CrossIcon2.default, { className: _Tooltip2.default['crossIcon--' + color] })
                    )
                )
            )
        );
    };

    return Tooltip;
}(_preact.Component);

exports.default = Tooltip;