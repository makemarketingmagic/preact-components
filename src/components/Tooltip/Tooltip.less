@import '../common/mixins.less';
@import '../common/variables.less';

.tooltip {
    background: @white;
    border-color: @white;
    border-radius: 4px;
    box-shadow: 0 4px 8px 0 rgba(0,0,0,0.04), 0 1px 2px 2px rgba(0,0,0,0.08);
    padding: 16px 32px;
    display: flex;
}

.tooltip::before {
    content: "";
    box-shadow: 10px -10px 0 10px #FFF, 0 4px 8px 0 rgba(0,0,0,.04), 0 1px 2px 2px rgba(0,0,0,.08);
    align-self: center;
    width: 16px;
    height: 16px;
    display: block;
    border-radius: 4px;
    transform: rotate(45deg);
    background: @white;
    position: relative;
    left: -39px;
    top: 4px;
    .responsive(600px, {
        display: none;
    });
}

.tooltipName {
    font-size: 14px;
    line-height: 17px;
    font-weight: 600;
}

.tooltipPosition {
    font-size: 12px;
    line-height: 14px;
    text-transform: uppercase;
}

.tooltipPosition--red {
    color: @red;
}

.tooltipPosition--pink {
    color: @pink;
}

.tooltipPosition--orange {
    color: @orange;
}

.tooltipPosition--yellow {
    color: @yellow;
}

.tooltipPosition--green {
    color: @green;
}

.tooltipPosition--grey {
    color: @grey; 
}

.tooltipPosition--blue {
    color: @blue;
}

.tooltipMessage {
    font-size: 12px;
    line-height: 14px;
    margin-top: 8px;
}

.tooltipContainer {
    display: flex;
    flex-direction: column;
    justify-content: center;
    transform-origin: left center;
    transition: transform 150ms ease-in;
}

.tooltipContainerOpen {
    transform: scaleX(1);
}

.tooltipContainerClosed {
    transform: scaleX(0);
}

.tooltipContents {
    display: flex;
    flex-direction: row;
}

.closeButton {
    flex: 0 0 8px;
    height: 24px;
    cursor: pointer;
}

.crossIcon--red {
    fill: @red;
}

.crossIcon--pink {
    fill: @pink;
}

.crossIcon--orange {
    fill: @orange;
}

.crossIcon--yellow {
    fill: @yellow;
}

.crossIcon--green {
    fill: @green;
}

.crossIcon--grey {
    fill: @grey; 
}

.crossIcon--blue {
    fill: @blue;
}
