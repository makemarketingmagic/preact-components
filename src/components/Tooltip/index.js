import { h, Component } from 'preact';
import CrossIcon from '../icons/CrossIcon';
import styled from 'styled-components';
import { colors, media } from '../common/scMixins';
import { transparentize } from 'polished';

const TooltipEl = styled.div`
    background: ${colors.white};
    border-color: ${colors.white};
    border-radius: 4px;
    box-shadow: 0 4px 8px 0 ${transparentize(0.96, colors.black)}, 0 1px 2px 2px ${transparentize(0.92, colors.black)};
    padding: 16px 32px;
    display: flex;
    &::before {
        content: '';
        box-shadow: 10px -10px 0 10px ${colors.white}, 0 4px 8px 0 ${transparentize(0.96, colors.black)}, 0 1px 2px 2px ${transparentize(0.92, colors.black)};
        align-self: center;
        width: 16px;
        height: 16px;
        display: block;
        border-radius: 4px;
        transform: rotate(45deg);
        background: ${colors.white};
        position: relative;
        left: -39px;
        top: 4px;
        ${media.mobile`
            display: none;
        `}
    }
`

const TooltipName = styled.div`
    font-size: 14px;
    line-height: 17px;
    font-weight: 600;
`

const TooltipPosition = styled.div`
    font-size: 12px;
    line-height: 14px;
    text-transform: uppercase;
    color: ${props => colors[props.color] || colors.black};
`

const TooltipMessage = styled.div`
    font-size: 12px;
    line-height: 14px;
    margin-top: 8px;
`

const TooltipContainer = styled.div`
    display: flex;
    flex-direction: column;
    justify-content: center;
    transform-origin: left center;
    transition: transform 150ms ease-in;
    transform: ${props => props.open ? 'scaleX(1)' : 'scaleX(0)'};
`

const TooltipContents = styled.div`
    display: flex;
    flex-direction: row;
`

const CloseButton = styled.div`
    flex: 0 0 8px;
    height: 24px;
    cursor: pointer;
`

const CrossIconEl = styled(CrossIcon)`
    fill: ${props => colors[props.color] || colors.black};
`

export default class Tooltip extends Component {
    constructor(props) {
        super(props)
        this.state = {
            visible: props.visible || true
        }
    }

    toggleTooltip = () => {
        this.setState({ visible: !this.state.visible })
    }


    render() {
        const { name, position, message, color, closable = true } = this.props;
        return (
            <TooltipContainer open={this.state.visible} class={'notranslate'}>
                <TooltipEl>
                    <TooltipContents>
                        <div>
                            <TooltipName>{name}</TooltipName>
                            <TooltipPosition color={color}>{position}</TooltipPosition>
                            <TooltipMessage>{message}</TooltipMessage>
                        </div>
                        {closable && <CloseButton onClick={this.toggleTooltip}>
                            <CrossIconEl color={color} />
                        </CloseButton>}
                    </TooltipContents>
                </TooltipEl>
            </TooltipContainer>
        )
    }
}