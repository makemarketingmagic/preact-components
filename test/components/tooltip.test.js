import { h } from 'preact';
import render from 'preact-render-to-string';
import { render as mock, fireEvent, debounceRenderingOff, wait } from 'preact-testing-library'
import Tooltip from './../../src/components/Tooltip/index';
import 'jest-dom/extend-expect'

describe('Tooltip', () => {
    it('should render container', () => {
        const tooltip = render(<Tooltip
            name='Test Name'
            position='Test Position'
            message='Test Message'
            color='orange'
        />)

        expect(tooltip).toMatchSnapshot()
    });

    it('should collapse on click', async () => {
        debounceRenderingOff()
        const tooltip = mock(<Tooltip
            name='Test Name'
            position='Test Position'
            message='Test Message'
            color='orange'
        />)
        fireEvent(
            tooltip.container.querySelector('.closeButton'),
            new MouseEvent('click', {
                bubbles: true,
                cancelable: true
            })
        )
        await wait(() => (expect(tooltip.getByText('Test Name')).not.toBeVisible()), {
            timeout: 2500
        })
        expect(tooltip.getByText('Test Name')).not.toBeVisible()
    })
})