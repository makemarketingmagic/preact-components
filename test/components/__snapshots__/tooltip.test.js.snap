// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip should render container 1`] = `
<div class="tooltipContainer notranslate tooltipContainerOpen">
  <div class="tooltip">
    <div class="tooltipContents">
      <div>
        <div class="tooltipName">Test Name</div>
        <div class="tooltipPosition tooltipPosition--orange">Test Position</div>
        <div class="tooltipMessage">Test Message</div>
      </div>
      <div class="closeButton">
        <svg width="8" height="8">
          <path class="crossIcon--orange" fill-rule="nonzero" d="M4 2.586L6.293.293a1 1 0 0 1 1.414 1.414L5.414 4l2.293 2.293a1 1 0 0 1-1.414 1.414L4 5.414 1.707 7.707A1 1 0 0 1 .293 6.293L2.586 4 .293 1.707A1 1 0 0 1 1.707.293L4 2.586z"></path>
        </svg>
      </div>
    </div>
  </div>
</div>
`;
